go-controller/pkg/ovn/base_network_controller_pods_test.go

```diff
@@ -0,0 +1,225 @@
+package ovn
+
+import (
+	"net"
+	"testing"
+	"time"
+
+	"github.com/onsi/gomega"
+	ovntest "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/testing"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+func TestBaseNetworkController_trackReleasedPodsBeforeStartup(t *testing.T) {
+	tests := []struct {
+		name           string
+		podAnnotations map[*corev1.Pod]map[string]*util.PodAnnotation
+		expected       map[string]sets.Set[string]
+	}{
+		{
+			name: "a scheduled/running annotated pod should not be considered released",
+			podAnnotations: map[*corev1.Pod]map[string]*util.PodAnnotation{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						UID: "running",
+					},
+				}: {
+					"default": {
+						IPs: []*net.IPNet{ovntest.MustParseIPNet("192.168.0.1/24")},
+					},
+				},
+			},
+			expected: map[string]sets.Set[string]{},
+		},
+		{
+			name: "a completed annotated pod should not be considered released",
+			podAnnotations: map[*corev1.Pod]map[string]*util.PodAnnotation{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						UID: "running",
+					},
+					Status: corev1.PodStatus{
+						Phase: corev1.PodSucceeded,
+					},
+				}: {
+					"default": {
+						IPs: []*net.IPNet{ovntest.MustParseIPNet("192.168.0.1/24")},
+					},
+				},
+			},
+			expected: map[string]sets.Set[string]{},
+		},
+		{
+			name: "a completed pod sharing at least one IP with a running Pod should be considered released",
+			podAnnotations: map[*corev1.Pod]map[string]*util.PodAnnotation{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						UID: "completed",
+					},
+					Status: corev1.PodStatus{
+						Phase: corev1.PodSucceeded,
+					},
+				}: {
+					"default": {
+						IPs: []*net.IPNet{
+							ovntest.MustParseIPNet("192.168.0.1/24"),
+							ovntest.MustParseIPNet("fd11::1/64"),
+						},
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						UID: "running",
+					},
+				}: {
+					"default": {
+						IPs: []*net.IPNet{
+							ovntest.MustParseIPNet("192.168.0.2/24"),
```

Comment:
Yes, and they are tracked individually. You will see in `trackReleasedPodsBeforeStartup` that we don't make any distinction by family, we just track individual IPs.

Reply:
I am assuming though that the IPs are released in block as there is no error path that would release just one of the IPs of a pod and not the other (release of a single IP can't fail). But I guess it would be good to add a comment about this somewhere.

Reply:
I think in the past we have always assumed this to be the case and we allocate them in a block. We could never end up with ip inconsistency between families. Maybe we should just fix the test cases to be consistent with the ips to avoid confusion.

Reply:
Right but the allocation case carries far more assumptions. For example, that we always allocate IPs for IPv4 and IPv6 all the time, and that the IPv4 and IPv6 subnets have the same size; all of which are the case now but might no be the case in the future. It is easy enough to avoid making those assumptions so I went for it.

In any case this is unit testing `trackReleasedPodsBeforeStartup` that has that capability. I will add a comment on test clarifying this.